File: apps/backend/app/routers/outreach.py
"""Outreach endpoints for V2 parsing, analysis, generation, history, and templates."""

import logging
from typing import Any

from fastapi import APIRouter, HTTPException

from app.config_cache import get_content_language
from app.database import db
from app.schemas.models import (
    AnalysisJobInput,
    DedupedHistoryResponse,
    GroupDetailRequest,
    GroupDetailResponse,
    GroupsAnalysisRequest,
    GroupsAnalysisResponse,
    MatchAnalysisV2Request,
    MatchAnalysisV2Response,
    OutreachGenerateV2Request,
    OutreachGenerateV2Response,
    OutreachHistoryResponse,
    OutreachRecordResponse,
    OutreachTemplateResponse,
    ParseJobRequest,
    ParsedJobInfo,
    SaveTemplateRequest,
)
from app.services import job_parser, match_analysis, outreach_generator
from app.services import outreach_history_analysis as history_analysis_svc


router = APIRouter(prefix="/outreach", tags=["Outreach"])
logger = logging.getLogger(__name__)


def _get_processed_resume_data(resume_id: str) -> dict[str, Any]:
    """Fetch structured resume data for standalone outreach features."""
    resume = db.get_resume(resume_id)
    if not resume:
        raise HTTPException(status_code=404, detail="Resume not found")

    processed_data = resume.get("processed_data")
    if not processed_data:
        raise HTTPException(
            status_code=409,
            detail="Resume has no processed data and is not ready for outreach generation.",
        )

    return processed_data


@router.post("/parse-job", response_model=ParsedJobInfo)
async def parse_job(request: ParseJobRequest) -> ParsedJobInfo:
    """Parse raw job description text into structured data via LLM."""
    try:
        return await job_parser.parse_job_info(request.job_info)
    except Exception as exc:
        logger.error("Job parsing failed: %s", exc)
        raise HTTPException(
            status_code=500,
            detail="Failed to parse job info. Please try again.",
        ) from exc


@router.post("/match-analysis", response_model=MatchAnalysisV2Response)
async def generate_match_analysis(
    request: MatchAnalysisV2Request,
) -> MatchAnalysisV2Response:
    """Analyze resume/job fit using structured job info."""
    processed_data = _get_processed_resume_data(request.resume_id)
    language = get_content_language()
    try:
        analysis = await match_analysis.analyze_match_v2(
            processed_data,
            request.parsed_job,
            language=language,
        )
        return MatchAnalysisV2Response(**analysis)
    except Exception as exc:
        logger.error("Match analysis V2 failed: %s", exc)
        raise HTTPException(
            status_code=500,
            detail="Failed to analyze match. Please try again.",
        ) from exc


@router.post("/generate", response_model=OutreachGenerateV2Response)
async def generate_outreach(
    request: OutreachGenerateV2Request,
) -> OutreachGenerateV2Response:
    """Generate a template-based outreach message and persist it to history."""
    processed_data = _get_processed_resume_data(request.resume_id)
    language = get_content_language()

    try:
        message = await outreach_generator.generate_outreach_message(
            processed_data,
            request.parsed_job,
            request.match_analysis,
            request.template,
            language=language,
        )
    except Exception as exc:
        logger.error("Outreach generation failed: %s", exc)
        raise HTTPException(
            status_code=500,
            detail="Failed to generate message. Please try again.",
        ) from exc

    record = db.create_outreach_record(
        resume_id=request.resume_id,
        parsed_job=request.parsed_job.model_dump(),
        match_analysis=request.match_analysis.model_dump(),
        generated_message=message,
        template_snapshot=request.template.model_dump(),
    )
    return OutreachGenerateV2Response(message=message, record_id=record["record_id"])


@router.get("/history", response_model=OutreachHistoryResponse)
def list_history() -> OutreachHistoryResponse:
    """Return all outreach history records, newest first."""
    records = db.list_outreach_records()
    return OutreachHistoryResponse(records=[OutreachRecordResponse(**record) for record in records])


@router.delete("/history/{record_id}")
def delete_history_record(record_id: str) -> dict[str, str]:
    """Delete a history record by ID."""
    deleted = db.delete_outreach_record(record_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Record not found")
    return {"message": "Record deleted"}


@router.get("/template", response_model=OutreachTemplateResponse)
def get_template() -> OutreachTemplateResponse:
    """Return the saved template or default values."""
    template = db.get_outreach_template()
    if template:
        return OutreachTemplateResponse(**template)

    return OutreachTemplateResponse(**db.DEFAULT_OUTREACH_TEMPLATE)


@router.put("/template", response_model=OutreachTemplateResponse)
def save_template(request: SaveTemplateRequest) -> OutreachTemplateResponse:
    """Save or update the single active outreach template."""
    template = db.save_outreach_template(
        opening=request.opening,
        middle_rules=request.middle_rules,
        closing=request.closing,
    )
    return OutreachTemplateResponse(**template)


# ---------------------------------------------------------------------------
# History Analysis Endpoints
# ---------------------------------------------------------------------------


@router.get("/history/deduped", response_model=DedupedHistoryResponse)
def get_deduped_history() -> DedupedHistoryResponse:
    """Return deduplicated outreach history for the analysis page.

    Deduplication is performed on company_name + job_title (normalized).
    Records with missing fields are returned in the unparsed_jobs bucket.
    No LLM is called.
    """
    records = db.list_all_outreach_records()
    jobs, unparsed_jobs = history_analysis_svc.dedup_outreach_records(records)
    return DedupedHistoryResponse(jobs=jobs, unparsed_jobs=unparsed_jobs)


@router.post("/history-analysis/groups", response_model=GroupsAnalysisResponse)
async def analyze_groups(request: GroupsAnalysisRequest) -> GroupsAnalysisResponse:
    """Stage-1 analysis: group deduplicated jobs into role types via LLM.

    The request body must contain the same deduplicated job list currently
    held by the frontend page snapshot (from GET /history/deduped).
    Duplicate job_keys in the request are rejected with HTTP 400.
    """
    # Uniqueness constraint: job_keys must not repeat within the request
    seen: set[str] = set()
    for job in request.jobs:
        if job.job_key in seen:
            raise HTTPException(
                status_code=400,
                detail=f"Duplicate job_key in request: {job.job_key}",
            )
        seen.add(job.job_key)

    try:
        return await history_analysis_svc.analyze_job_groups(
            request.jobs, language=request.content_language
        )
    except Exception as exc:
        logger.error("Stage-1 history analysis failed: %s", exc)
        raise HTTPException(
            status_code=500,
            detail="Failed to analyze job groups. Please try again.",
        ) from exc


@router.post("/history-analysis/group-detail", response_model=GroupDetailResponse)
async def analyze_group_detail(request: GroupDetailRequest) -> GroupDetailResponse:
    """Stage-2 analysis: extract common traits for a specific role type via LLM.

    The jobs in the request body must be pre-filtered by the frontend
    (using group.job_keys against the page snapshot).
    """
    try:
        return await history_analysis_svc.analyze_group_detail(
            group_name=request.group_name,
            jobs=request.jobs,
            language=request.content_language,
        )
    except Exception as exc:
        logger.error("Stage-2 history analysis failed: %s", exc)
        raise HTTPException(
            status_code=500,
            detail="Failed to analyze group detail. Please try again.",
        ) from exc

File: apps/frontend/lib/api/outreach.ts
import { apiDelete, apiFetch, apiPost, apiPut } from './client';
import type { ResumeListItem } from './resume';

export type OutreachResumeListItem = ResumeListItem;

export interface ParsedJobInfo {
  company_name: string;
  company_info: string;
  job_title: string;
  job_description: string;
  requirements: string[];
  bonus_items: string[];
  required_materials: string[];
  raw_text: string;
}

export interface OutreachMatchAnalysisV2 {
  match_percentage: number;
  requirement_matches: string[];
  requirement_gaps: string[];
  bonus_matches: string[];
  bonus_gaps: string[];
  strengths: string[];
  weaknesses: string[];
  analysis_summary: string;
}

export interface OutreachTemplateData {
  template_id?: string;
  opening: string;
  middle_rules: string;
  closing: string;
  created_at?: string;
  updated_at?: string;
}

export interface OutreachGenerateResult {
  message: string;
  record_id: string;
}

export interface OutreachRecord {
  record_id: string;
  resume_id: string;
  parsed_job: ParsedJobInfo;
  match_analysis: OutreachMatchAnalysisV2;
  generated_message: string;
  template_snapshot: OutreachTemplateData;
  created_at: string;
}

async function readError(response: Response, fallback: string): Promise<string> {
  const raw = await response.text().catch(() => '');
  if (!raw) return fallback;

  try {
    const parsed = JSON.parse(raw) as { detail?: string };
    if (typeof parsed.detail === 'string' && parsed.detail.trim()) {
      return parsed.detail;
    }
  } catch {
    // Fall back to raw text.
  }

  return raw;
}

export async function fetchOutreachResumeList(): Promise<OutreachResumeListItem[]> {
  const response = await apiFetch('/resumes/list?include_master=true');
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to load resumes (status ${response.status}).`)
    );
  }

  const payload = (await response.json()) as { data: OutreachResumeListItem[] };
  return payload.data;
}

export async function parseJobInfo(jobInfo: string): Promise<ParsedJobInfo> {
  const response = await apiPost('/outreach/parse-job', { job_info: jobInfo });
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to parse job info (status ${response.status}).`)
    );
  }

  return (await response.json()) as ParsedJobInfo;
}

export async function generateMatchAnalysis(
  resumeId: string,
  parsedJob: ParsedJobInfo
): Promise<OutreachMatchAnalysisV2> {
  const response = await apiPost('/outreach/match-analysis', {
    resume_id: resumeId,
    parsed_job: parsedJob,
  });
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to analyze match (status ${response.status}).`)
    );
  }

  return (await response.json()) as OutreachMatchAnalysisV2;
}

export async function generateOutreachMessage(
  resumeId: string,
  parsedJob: ParsedJobInfo,
  matchAnalysis: OutreachMatchAnalysisV2,
  template: OutreachTemplateData
): Promise<OutreachGenerateResult> {
  const response = await apiPost('/outreach/generate', {
    resume_id: resumeId,
    parsed_job: parsedJob,
    match_analysis: matchAnalysis,
    template: {
      opening: template.opening,
      middle_rules: template.middle_rules,
      closing: template.closing,
    },
  });
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to generate message (status ${response.status}).`)
    );
  }

  return (await response.json()) as OutreachGenerateResult;
}

export async function fetchOutreachHistory(): Promise<OutreachRecord[]> {
  const response = await apiFetch('/outreach/history');
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to load history (status ${response.status}).`)
    );
  }

  const payload = (await response.json()) as { records: OutreachRecord[] };
  return payload.records;
}

export async function deleteOutreachRecord(recordId: string): Promise<void> {
  const response = await apiDelete(`/outreach/history/${recordId}`);
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to delete record (status ${response.status}).`)
    );
  }
}

export async function fetchOutreachTemplate(): Promise<OutreachTemplateData> {
  const response = await apiFetch('/outreach/template');
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to load template (status ${response.status}).`)
    );
  }

  return (await response.json()) as OutreachTemplateData;
}

export async function saveOutreachTemplate(
  template: OutreachTemplateData
): Promise<OutreachTemplateData> {
  const response = await apiPut('/outreach/template', {
    opening: template.opening,
    middle_rules: template.middle_rules,
    closing: template.closing,
  });
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to save template (status ${response.status}).`)
    );
  }

  return (await response.json()) as OutreachTemplateData;
}

// ---------------------------------------------------------------------------
// History Analysis Types
// ---------------------------------------------------------------------------

export interface DedupedJob {
  job_key: string;
  sample_record_id: string;
  company_name: string;
  job_title: string;
  job_description: string;
  requirements: string[];
  bonus_items: string[];
  duplicate_count: number;
  latest_applied_at: string;
}

export interface UnparsedJob {
  record_id: string;
  company_name: string;
  job_title: string;
  raw_text_preview: string;
  created_at: string;
}

export interface DedupedHistoryData {
  jobs: DedupedJob[];
  unparsed_jobs: UnparsedJob[];
}

export interface JobGroup {
  group_id: string;
  group_name: string;
  summary: string;
  job_keys: string[];
  sample_titles: string[];
  count: number;
}

export interface GroupsAnalysisResult {
  analyzed_job_count: number;
  truncated_from_total: number;
  groups: JobGroup[];
  uncategorized_job_keys: string[];
}

export interface GroupDetailResult {
  group_name: string;
  analysis_summary: string;
  technology_stack: string[];
  common_requirements: string[];
  common_bonus_points: string[];
  common_responsibilities: string[];
  representative_titles: string[];
}

// ---------------------------------------------------------------------------
// History Analysis API Functions
// ---------------------------------------------------------------------------

export async function fetchDedupedHistory(): Promise<DedupedHistoryData> {
  const response = await apiFetch('/outreach/history/deduped');
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to load history (status ${response.status}).`)
    );
  }
  return (await response.json()) as DedupedHistoryData;
}

export async function analyzeJobGroups(
  jobs: DedupedJob[],
  contentLanguage: string
): Promise<GroupsAnalysisResult> {
  const response = await apiPost('/outreach/history-analysis/groups', {
    content_language: contentLanguage,
    jobs,
  });
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to analyze job groups (status ${response.status}).`)
    );
  }
  return (await response.json()) as GroupsAnalysisResult;
}

export async function analyzeGroupDetail(
  groupName: string,
  jobs: DedupedJob[],
  contentLanguage: string
): Promise<GroupDetailResult> {
  const response = await apiPost('/outreach/history-analysis/group-detail', {
    group_name: groupName,
    content_language: contentLanguage,
    jobs,
  });
  if (!response.ok) {
    throw new Error(
      await readError(response, `Failed to analyze group detail (status ${response.status}).`)
    );
  }
  return (await response.json()) as GroupDetailResult;
}
